Link decoration filtering should be able to be conditional on URL path

charliewolfe opened the pull request:
#### e375751b5b526c89d630ddccad2e4f2df3ac0d2a
<pre>
Link decoration filtering should be able to be conditional on URL path
<a href="https://bugs.webkit.org/show_bug.cgi?id=266382">https://bugs.webkit.org/show_bug.cgi?id=266382</a>
<a href="https://rdar.apple.com/111189162">rdar://111189162</a>

Reviewed by Wenson Hsieh.

There are situations where we want query parameter removal to only happen for particular URL paths. This
patch adds support to do so.

Since link decoration filtering is now conditional on domains and paths, I removed
`m_domainScopedLinkDecorationFilteringData` and made `m_linkDecorationFilteringData` map from query
parameter to the domain and path conditionals.

I also made changes so that our API tests support testing link decoration filtering by domain and path.

* Source/WebCore/PAL/pal/spi/cocoa/WebPrivacySPI.h:
* Source/WebCore/page/LinkDecorationFilteringData.h:
(WebCore::LinkDecorationFilteringData::LinkDecorationFilteringData):
(WebCore::LinkDecorationFilteringData::operator=):
(WebCore::LinkDecorationFilteringData::linkDecoration): Deleted.
* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
(WebKit::LinkDecorationFilteringController::updateStrings):
(WebKit::requestLinkDecorationFilteringData):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::applyLinkDecorationFilteringWithResult):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setLinkDecorationFilteringData):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Tools/TestWebKitAPI/Tests/WebKit/AdvancedPrivacyProtections.mm:
(TestWebKitAPI::makeQueryParameterRequestHandler):
(TestWebKitAPI::QueryParameterRequestSwizzler::QueryParameterRequestSwizzler):
(TestWebKitAPI::QueryParameterRequestSwizzler::update):
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/272076@main">https://commits.webkit.org/272076@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/60ab0a86563d09c0060caee9cabc2f3016df8e67

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/30562 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/9244 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/32215 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/33059 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/27633 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/31266 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/11530 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/6496 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/27552 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/30870 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/7749 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/27371 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/6637 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/6796 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/27226 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/34395 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/27819 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/27721 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/34395 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/6806 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/4921 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/34395 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/8542 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/7537 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3958 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/7361 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Tools/TestWebKitAPI/Tests/WebKit/AdvancedPrivacyProtections.mm:
```diff
@@ -99,11 +100,15 @@ - (id)callAsyncJavaScriptAndWait:(NSString *)script arguments:(NSDictionary *)ar
+        RunLoop::main().dispatch([&didHandleRequest, parameters = WTFMove(parameters), domains = WTFMove(domains), paths = WTFMove(paths), completion = makeBlockPtr(completion)]() mutable {
+            auto data = adoptNS([PAL::allocWPLinkFilteringDataInstance() initWithQueryParameters:parameters.get()
+#if defined(HAS_WEB_PRIVACY_LINK_FILTERING_RULE_PATH)
+            domains:domains.get() paths:paths.get()
```
Nit - this should be indented one more level.

whsieh commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -8876,20 +8876,26 @@ bool WebPage::isUsingUISideCompositing() const
-    for (auto& data : strings) {
-        if (data.domain.isEmpty()) {
-            m_linkDecorationFilteringData.add(data.linkDecoration);
+    for (const auto& data : strings) {
```
Nit - I believe we generally prefer `auto&` over `const auto&`, even if `data` isn't being mutated.

charliewolfe commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -768,10 +768,19 @@
         return { url, DidFilterLinkDecoration::No };
 
     auto sanitizedURL = url;
+    auto removedParameters = WTF::removeQueryParameters(sanitizedURL, [&](const auto& parameter) {
```
I'll also change this to `auto&` before landing.

charliewolfe commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -768,10 +768,19 @@
+        const auto& conditionals = it->value;
+        bool missingOrFoundDomain = conditionals.domains.isEmpty() || conditionals.domains.contains(RegistrableDomain { url });
+        bool missingOrFoundPath = conditionals.paths.isEmpty() || std::any_of(conditionals.paths.begin(), conditionals.paths.end(),
+            [&url](const auto& path) {
```
Here too.

sysrqb commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -768,10 +768,19 @@
+
+        const auto& conditionals = it->value;
+        bool missingOrFoundDomain = conditionals.domains.isEmpty() || conditionals.domains.contains(RegistrableDomain { url });
+        bool missingOrFoundPath = conditionals.paths.isEmpty() || std::any_of(conditionals.paths.begin(), conditionals.paths.end(),
```
Nit: something like `unconditional`/`unrestricted`/`empty` might be clearer (e.g., `unconditionalOrFoundDomain`). To me `missing` implies that something is potentially wrong.

charliewolfe replied:
👍 I’ll change this to `emptyOrFoundDomain`.

cdumez replied:
That's not a great name either. Per coding style, boolean variables in WebKit need to have a prefix such as 'is' / 'has'.

charliewolfe replied:
Ok, I'll use `isEmptyOrFoundDomain`.

sysrqb commented on Source/WebCore/PAL/pal/spi/cocoa/WebPrivacySPI.h:
```diff
@@ -127,6 +128,12 @@ typedef void (^WPStorageAccessUserAgentStringQuirksDataCompletionHandler)(WPStor
 @end
 #endif
 
+#if !defined(HAS_WEB_PRIVACY_LINK_FILTERING_RULE_PATH) && HAVE(WEB_PRIVACY_FRAMEWORK)
```
Is the `HAVE(WEB_PRIVACY_FRAMEWORK)` important here?

charliewolfe replied:
Maybe not, but whenever `HAVE(WEB_PRIVACY_FRAMEWORK)` is false this doesn’t need to be compiled, because `WPLinkFilteringRule` would be declared above.

sysrqb commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -8876,20 +8876,26 @@ bool WebPage::isUsingUISideCompositing() const
-        if (data.domain.isEmpty()) {
-            m_linkDecorationFilteringData.add(data.linkDecoration);
+        if (!m_linkDecorationFilteringData.isValidKey(data.linkDecoration)) {
+            ASSERT_NOT_REACHED();
```
Maybe add a release log for this, too. We'll only crash with the assertion in a pretty specific case.

sysrqb commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -8876,20 +8876,26 @@ bool WebPage::isUsingUISideCompositing() const
+            if (auto& domains = it->value.domains; domains.isValidValue(data.domain))
+                domains.add(data.domain);
+            else
+                ASSERT_NOT_REACHED();
```
And what's the correct behavior here? Should we just skip the domain and continue on to checking the path, or should we consider the data corrupt and do something else?

charliewolfe replied:
I guess we could early return here, I don't have a preference. I just wanted these assertions to be hit if something was misconfigured when running the API tests.

sysrqb reviewed (approved):
LGTM with nits

charliewolfe committed:
Link decoration filtering should be able to be conditional on URL path
https://bugs.webkit.org/show_bug.cgi?id=266382
rdar://111189162

Reviewed by Wenson Hsieh.

There are situations where we want query parameter removal to only happen for particular URL paths. This
patch adds support to do so.

Since link decoration filtering is now conditional on domains and paths, I removed
`m_domainScopedLinkDecorationFilteringData` and made `m_linkDecorationFilteringData` map from query
parameter to the domain and path conditionals.

I also made changes so that our API tests support testing link decoration filtering by domain and path.

* Source/WebCore/PAL/pal/spi/cocoa/WebPrivacySPI.h:
* Source/WebCore/page/LinkDecorationFilteringData.h:
(WebCore::LinkDecorationFilteringData::LinkDecorationFilteringData):
(WebCore::LinkDecorationFilteringData::operator=):
(WebCore::LinkDecorationFilteringData::linkDecoration): Deleted.
* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
(WebKit::LinkDecorationFilteringController::updateStrings):
(WebKit::requestLinkDecorationFilteringData):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::applyLinkDecorationFilteringWithResult):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setLinkDecorationFilteringData):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Tools/TestWebKitAPI/Tests/WebKit/AdvancedPrivacyProtections.mm:
(TestWebKitAPI::makeQueryParameterRequestHandler):
(TestWebKitAPI::QueryParameterRequestSwizzler::QueryParameterRequestSwizzler):
(TestWebKitAPI::QueryParameterRequestSwizzler::update):
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/272076@main
